### PR TITLE
A Fix for the permalinks in #1136

### PR DESCRIPTION
**Pull request by aslowwriter**
## Code changes
When taking another look at my PR #1136 I noticed that it didn't actually solve the problem because it didn't merge the permalinks properly. apologies for that. I checked that Also, my git fu is not amazing but I've tried to keep the commit history as clean as possible, I hope it's okay like this. 

## Checklist
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/getzola/zola/pulls) for the same update/change?
* [x] Are you doing the PR on the `next` branch?
* [x] Have you created/updated the relevant documentation page(s)?

### CODE REVIEW

**Comment by Keats** [comment]
The test failures seem legit?

**Comment by aslowwriter** [comment]
they were, I had forgotten to run the entire test suite rather than just the ones for the taxonomy, that's my bad. The point is that the taxonomy names now get saved slightly differently to detect the collisions where necessary. The only change necessary was to add `| replace(from="-", to=" ") |  title ` to the relevant templates where `tag.name` was used, it should be good now

File: components/templates/src/builtins/atom.xml
Comment (Keats) [inline components/templates/src/builtins/atom.xml]: sadly that doesn't work if you have unicode :/
The deduplication should happen without needing any changes to the templates

**Review by Keats — CHANGES_REQUESTED** [review]

File: components/templates/src/builtins/atom.xml
Comment (aslowwriter) [reply components/templates/src/builtins/atom.xml]: okay I'll have a look and see what I can do

**Comment by Keats** [comment]
Can you do a separate PR with some failing tests for the permalinks? I can have a look as well.

**Comment by aslowwriter** [comment]
I think I found a solution that works and doesn't require moddifications to the templates to pass all tests. If you're happy about the changes I'll try and clean up the history before you merge the request

File: components/library/src/taxonomies/mod.rs
Comment (Keats) [inline components/library/src/taxonomies/mod.rs]: if you merge you can pass `Self` I think

File: components/library/src/taxonomies/mod.rs
Comment (Keats) [inline components/library/src/taxonomies/mod.rs]: stable

File: components/library/src/taxonomies/mod.rs
Comment (Keats) [inline components/library/src/taxonomies/mod.rs]: can be removed?

**Review by Keats — CHANGES_REQUESTED** [review]
Looks ok to me. Do you think there are enough tests covering that or should we add more?

File: components/library/src/taxonomies/mod.rs
Comment (aslowwriter) [reply components/library/src/taxonomies/mod.rs]: noted will apply this shortly

File: components/library/src/taxonomies/mod.rs
Comment (aslowwriter) [reply components/library/src/taxonomies/mod.rs]: I assume you mean that you prefer the original bit since that is already stable, which is fine, I'll revert it when I make the change

Comment (aslowwriter) [reply components/library/src/taxonomies/mod.rs]: Edit, sorry I just realised sorting by slug first is actually necessary since the deduping otherwise fails. Tests fail with the original sorting

File: components/library/src/taxonomies/mod.rs
Comment (aslowwriter) [reply components/library/src/taxonomies/mod.rs]: I thought this would cause ownership conflicts but after trying it out you are actually correct. Changed this in the next commit

**Comment by aslowwriter** [comment]
> Looks ok to me. Do you think there are enough tests covering that or should we add more?

Thanks for the feedback. I added tests for all cases that I could think of that would result in the same permalink (added whitespace, inconsistent capitalisation and substituting spaces for hyphens) so unless you can think of something on top of that I think it should be okay? 

The only thing that is perhaps left untested is how this all interacts with Unicode, but I thought that valid URLs are limited to ASCII, so I didn't add anything for that. I don't actually know enough to make a definitive comment on that so I'll defer to your judgement on that

**Comment by Keats** [comment]
> The only thing that is perhaps left untested is how this all interacts with Unicode, but I thought that valid URLs are limited to ASCII, so I didn't add anything for that.

They get escaped but you can get URL outside of ASCII: https://fr.wikipedia.org/wiki/%C3%89cole (eg École).
If someone is not turning paths into slugs, the 2 taxonomies ecole and école should NOT be merged. If you are slugifying them on the other hand, they should end up merged.

**Comment by aslowwriter** [comment]
I've added two tests to cover those two cases

**Review by Keats — APPROVED** [review]

**Comment by Keats** [comment]
Thanks!

**Pull request merged**
